### DISCUSSION TITLE
MustardUI 2

### DISCUSSION THREAD

**Discussion by Mustard2** [Mustard2/MustardUI#5] [General]
Hi everyone!

I would like to announce a new version of MustardUI which will be released very soon, that in this discussion I will call MustardUI 2 (not the final name). This is a post to explain the main changes and what you might expect as a creator and/or user of the models which use this UI.

Before considering the changes, let me explain the main problems of MustardUI 1 (the old version).

- MustardUI1 was entirely based on the creation of new Blender properties for each property that was available in the UI. This approach creates a huge amount of work, since we have to manually create all the new properties into Blender, which can be very time consuming. Moreover, the creator or the user that wants to add a new property can not do it unless he/she knows where to look in the code, which means he/she should have a proper understanding of how Python/Blender Python work (in addition to decipher my bad way of coding!).
- In the script, a lot of important variables were stored as global variables. Moreover, the UI was specific for each model, and each of them had to use a specific script. This creates a problem when more characters are available in the scene. In fact, in MustardUI 1 only one character UI is supported: if 2 models are on the scene, only one can be controlled with the UI.
- I created the script with a few functionalities in mind, and added a lot of them with unconsistent programming conventions (not to mention the lack of proper comments). This further creates issues for other people when they read the code.

Note that I take full responsability for not fixing these problems while I was coding this first version. I got lazy and the prize was a gigantic coding mess which was impossible to fix.

The new version has been completely rewritten. The main new features/changes are the following.
- Although the main new features are more important for the creators rather than the users, which will see small to now differences with respect to MustardUI 1, the UI is now more flexible and **the users can add new properties/outfits etc.. without any modification of the code** (so no Python needed!). But this new feature does not stop on the user side, because **the new UI allows creators to quickly set up a UI with no need to write a single line of code** (not even the name of the character in the .py file). For instance, you can see some of the properties you will be able to modify in the pictures below.
![config_out](https://user-images.githubusercontent.com/60780948/117977240-e6fbdd00-b330-11eb-94b8-540d07f17e11.png) ![config_others](https://user-images.githubusercontent.com/60780948/117977238-e6634680-b330-11eb-81f5-806b33536097.png)
- To add new outfits, you will not need to write the name of the collections in the script (with all the problems related to copy-paste, etc..) anymore! While in the Configuration mode, **just right click on any collection to add it as an outfit**.
![add_outfit](https://user-images.githubusercontent.com/60780948/117977234-e5cab000-b330-11eb-99dd-67b612f37e58.PNG)
What happens if you delete/rename a collection? The addon will automatically take care of these problems, so no more nightmares if some outfits are deleted on the scene (for instance before renderings to save memory).
- **No more global variables**, and most of the functions are defined inside classes to avoid any incompatibility with other add-ons. Plus, everything is now consistently named, with a MustardUI in front of all properties created by the MustardUI add-on.
- **Support for more characters in the same scene** (I know, finally!).

Note that this is just a part of the work, more features has been and will be added.
I hope to use this UI very soon to show the changes I made. :)

P. S. This new version will be fully compatible with old models. The new configuration panel will automatically take care of everything. :)

**Comment by jfixit** [comment]
Looks awesome and I can't wait to try it out.

**Reply by Mustard2** [reply]
Already available: https://github.com/Mustard2/MustardUI/releases/tag/0.20.5 :)
